File: linkedlist/sll.c
#include <stdio.h>
#include <stdlib.h>

struct list {
    int data;
    struct list *next;
};

struct list *first = NULL, *last = NULL;

void insertAtBeg(int ele) {
    struct list *Newnode;
    Newnode = (struct list *)malloc(sizeof(struct list));
    Newnode->next = NULL;
    Newnode->data = ele;

    if (first == NULL) {
        first = last = Newnode;
    } else {
        Newnode->next = first;
        first = Newnode;
    }
}

void insertAtEnd(int ele) {
    struct list *Newnode;
    Newnode = (struct list *)malloc(sizeof(struct list));
    Newnode->next = NULL;
    Newnode->data = ele;

    if (first == NULL) {
        first = last = Newnode;
    } else {
        last->next = Newnode;
        last = Newnode;
    }
}

void insertAtPos(int ele, int pos) {
    struct list *Newnode, *temp;
    int i;

    if (pos == 1) {
        insertAtBeg(ele);
        return;
    }

    Newnode = (struct list *)malloc(sizeof(struct list));
    Newnode->data = ele;
    Newnode->next = NULL;

    if (first == NULL) {
        printf("List is empty. Inserting at position 1.\n");
        first = last = Newnode;
        return;
    }

    temp = first;
    for (i = 1; i < pos - 1 && temp != NULL; i++) {
        temp = temp->next;
    }

    if (temp == NULL) {
        printf("Invalid position.\n");
        free(Newnode);
        return;
    }

    Newnode->next = temp->next;
    temp->next = Newnode;
    if (Newnode->next == NULL) {
        last = Newnode;
    }
}

void deleteBeg() {
    struct list *temp;
    if (first == NULL) {
        printf("List is empty\n");
    } else {
        temp = first;
        first = first->next;
        if (first == NULL) {
            last = NULL;
        }
        free(temp);
    }
}

void deleteEnd() {
    struct list *temp;
    if (first == NULL) {
        printf("List is empty\n");
    } else if (first->next == NULL) {
        free(first);
        first = last = NULL;
    } else {
        temp = first;
        while (temp->next != last) {
            temp = temp->next;
        }
        free(last);
        last = temp;
        last->next = NULL;
    }
}

void deletePos(int pos) {
    struct list *temp, *tempp;
    int i;

    if (first == NULL) {
        printf("List is empty\n");
        return;
    }

    if (pos == 1) {
        deleteBeg();
        return;
    }

    temp = first;
    for (i = 1; i < pos - 1 && temp != NULL; i++) {
        temp = temp->next;
    }

    if (temp == NULL || temp->next == NULL) {
        printf("Invalid position\n");
        return;
    }

    tempp = temp->next;
    temp->next = tempp->next;

    if (tempp == last) {
        last = temp;
    }

    free(tempp);
}

void search(int s) {
    struct list *temp = first;
    int flag = 0;

    if (first == NULL) {
        printf("List is Empty\n");
        return;
    }

    while (temp != NULL) {
        if (temp->data == s) {
            flag = 1;
            printf("Element Found\n");
            return;
        }
        temp = temp->next;
    }

    if (flag == 0) {
        printf("Element Not Found!\n");
    }
}

void display() {
    struct list *temp;
    if (first == NULL)
        printf("EMPTY LIST!\n");
    else {
        temp = first;
        while (temp != NULL) {
            printf("%d -> ", temp->data);
            temp = temp->next;
        }
        printf("NULL\n");
    }
}

int main() {
    int ele, pos, choice;

    while (1) {
        printf("\n---------------Menu----------------\n");
        printf("Press 1 to insert at the beginning.\n");
        printf("Press 2 to insert at the end.\n");
        printf("Press 3 to insert at a specific position.\n");
        printf("Press 4 to delete from the beginning.\n");
        printf("Press 5 to delete from the end.\n");
        printf("Press 6 to delete from a specific position.\n");
        printf("Press 7 to search for an element.\n");
        printf("Press 8 to display the list.\n");
        printf("Press 9 to exit.\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter element to insert at beginning: ");
                scanf("%d", &ele);
                insertAtBeg(ele);
                break;
            case 2:
                printf("Enter element to insert at end: ");
                scanf("%d", &ele);
                insertAtEnd(ele);
                break;
            case 3:
                printf("Enter element and position to insert: ");
                scanf("%d%d", &ele, &pos);
                insertAtPos(ele, pos);
                break;
            case 4:
                deleteBeg();
                break;
            case 5:
                deleteEnd();
                break;
            case 6:
                printf("Enter position to delete: ");
                scanf("%d", &pos);
                deletePos(pos);
                break;
            case 7:
                printf("Enter element to search: ");
                scanf("%d", &ele);
                search(ele);
                break;
            case 8:
                display();
                break;
            case 9:
                printf("Exiting...\n");
                exit(0);
            default:
                printf("Invalid choice, please try again!!\n");
        }
    }

    return 0;
}

File: linkedlist/Sll.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.first = None
        self.last = None

    def insert_at_beg(self, ele):
        new_node = Node(ele)
        if self.first is None:
            self.first = self.last = new_node
        else:
            new_node.next = self.first
            self.first = new_node

    def insert_at_end(self, ele):
        new_node = Node(ele)
        if self.first is None:
            self.first = self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node

    def insert_at_pos(self, ele, pos):
        new_node = Node(ele)
        if pos == 1:
            self.insert_at_beg(ele)
            return

        if self.first is None:
            print("List is empty. Inserting at position 1.")
            self.first = self.last = new_node
            return

        temp = self.first
        i = 1
        while i < pos - 1 and temp is not None:
            temp = temp.next
            i += 1

        if temp is None:
            print("Invalid position.")
            return

        new_node.next = temp.next
        temp.next = new_node
        if new_node.next is None:
            self.last = new_node

    def delete_beg(self):
        if self.first is None:
            print("List is empty")
        else:
            temp = self.first
            self.first = self.first.next
            if self.first is None:
                self.last = None
            del temp

    def delete_end(self):
        if self.first is None:
            print("List is empty")
        elif self.first.next is None:
            del self.first
            self.first = self.last = None
        else:
            temp = self.first
            while temp.next != self.last:
                temp = temp.next
            del self.last
            self.last = temp
            self.last.next = None

    def delete_pos(self, pos):
        if self.first is None:
            print("List is empty")
            return

        if pos == 1:
            self.delete_beg()
            return

        temp = self.first
        i = 1
        while i < pos - 1 and temp is not None:
            temp = temp.next
            i += 1

        if temp is None or temp.next is None:
            print("Invalid position")
            return

        to_delete = temp.next
        temp.next = to_delete.next
        if to_delete == self.last:
            self.last = temp
        del to_delete

    def search(self, s):
        temp = self.first
        while temp is not None:
            if temp.data == s:
                print("Element Found")
                return
            temp = temp.next
        print("Element Not Found!")

    def display(self):
        if self.first is None:
            print("EMPTY LIST!")
        else:
            temp = self.first
            while temp:
                print(f"{temp.data} -> ", end='')
                temp = temp.next
            print("NULL")


def main():
    ll = LinkedList()

    while True:
        print("\n---------------Menu----------------")
        print("1. Insert at the beginning")
        print("2. Insert at the end")
        print("3. Insert at a specific position")
        print("4. Delete from the beginning")
        print("5. Delete from the end")
        print("6. Delete from a specific position")
        print("7. Search for an element")
        print("8. Display the list")
        print("9. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            ele = int(input("Enter element to insert at beginning: "))
            ll.insert_at_beg(ele)
        elif choice == '2':
            ele = int(input("Enter element to insert at end: "))
            ll.insert_at_end(ele)
        elif choice == '3':
            ele = int(input("Enter element: "))
            pos = int(input("Enter position: "))
            ll.insert_at_pos(ele, pos)
        elif choice == '4':
            ll.delete_beg()
        elif choice == '5':
            ll.delete_end()
        elif choice == '6':
            pos = int(input("Enter position to delete: "))
            ll.delete_pos(pos)
        elif choice == '7':
            ele = int(input("Enter element to search: "))
            ll.search(ele)
        elif choice == '8':
            ll.display()
        elif choice == '9':
            print("Exiting...")
            break
        else:
            print("Invalid choice, please try again!!")


if __name__ == "__main__":
    main()
